### ISSUE TITLE
Typescript classes within separate it functions pollute / interfer with each other?

### ISSUE THREAD

**Issue by oleersoy** [mochajs/mocha#3425]
I asked about this on [SO first](https://stackoverflow.com/questions/51006697/preventing-the-scaffolding-setup-in-one-mocha-it-function-from-interfering-wit) just to make sure I was not missing something.  I'm going to publish the entire test suite soon, so that it's easy to see what's happening.  Just reporting it in case this is a bug or not supposed to happen.

**Comment by oleersoy** [comment]
Actually this could be a good thing.  In general if we have a class definition it should be universally unique in the runtime, regardless of where it is defined.

I was duplicating some of the class definitions by copying and pasting, and since they were decorated, the running of the decorators produced state that was incorrect.  Specifically, since these are validation decorators, it produced duplicate array items, so the fact that Mocha, so if Mocha had somehow isolated the class definition, I would not have seen this effect.

So to summarize I would say that Mocha not isolating the scope of classes to specific it blocks is a good thing.

**Comment by oleersoy** [comment]
So I've refactored my project to import test class definitions from scaffolding files.  However now I would like to test for multiple instances of the same class definition at runtime.  It's possible that this would occur, but how do we define a class, without having it pollute all other test files?  I'll experiment with this a bit and post back ....

**Comment by oleersoy** [comment]
So here something that might be an issue.  Suppose we have a single test file with an import like this:

```
import { NoOptionsValid } from "@test/decorators/IsDefined";

describe()....
```

In this case the person running the test might be expecting that the only decorators that will fire are the ones on `NoOptionsValid`, but if the person is importing other decorated classes in other test files, those decorators will also fire...which could lead to some strange results.
